### PR TITLE
Add "attach storage namespace [to repository]" IAM action type

### LINKED ISSUE #2220: Support policies to limit lakefs users to repo operations for specific buckets/blobs
According to the documentation [Authorization](https://docs.lakefs.io/reference/authorization.html) It seems like we are not able to limit the scope for lakefs users to a specific bucket/blob. 

## Use case 

We require to set up a multi-tenant data development environment such that teams are limited to managing repositories backed by a specific blob/bucket. This includes the create, delete repo operations only for repos that reside in a given bucket x. This would enable better encapsulation for data teams and make it easier to manage the stores from the cloud providers perspective as well. 

Using AWS as an example, I believe that being able to `lakectl auth users policies attach`  a functional policy like below to a lakefs user would suffice: 

```json 
{
  "Action": [
    "fs:*"
  ],
  "Effect": "Allow",
  "Resource": ["arn:aws:s3:::<BUCKET_NAME>", "arn:aws:s3:::<BUCKET_NAME_WITH_PATH_PREFIX>/*"]
}
```

Maybe, this could be generalised by providing an `arn` for the storage namespace: `arn:lakefs:fs:::storage/mystorage/*`.  If it's so that the mapping between storage namespace and repo is not 1-1, rather One To Many.

### PR DESCRIPTION
**Pull request by arielshaqed**
Fixes #2220 (and see discussion there).

### CODE REVIEW

**Comment by arielshaqed** [comment]
@ozkatz,  following our chats here's a simple implementation which I believe implements the solution that you outlined in #2246.  Is this what you intended?

**Review by ozkatz — CHANGES_REQUESTED** [review]
LGTM! I'm missing 2 important parts though:
1. Builtin policies that allow creating a repository should now include this new permission with a wildcard to preserve existing behavior
2. a data migration such that any policy in the user's database that included createRepository should also get this new permission with a wildcard.

**Comment by arielshaqed** [comment]
> LGTM! I'm missing 2 important parts though:
> 
>     1. Builtin policies that allow creating a repository should now include this new permission with a wildcard to preserve existing behavior

AFAIU this is _not_ required, we already have policies for actions `fs:*` that include ARNs `*`, and no specific policy applies to `CreateRepository`.

>     2. a data migration such that any policy in the user's database that included createRepository should also get this new permission with a wildcard.

True.  I believe that we can do this by adding a new rule allowing action `fs:AttachRepositoryNamespace` on `*` for any policy that allows `fs:CreateRepository`.  I think that I should add this with a check that it's not already been added, and then _do nothing on downgrade_: it is safer to keep the old permissions than to lose them silently.

**Comment by arielshaqed** [comment]
Working on upgrade script.  There's a split between possibilities here, complexity of SQL balanced against code complexity for a script that should run only once.
I would like the script to "add[...] a new rule allowing action fs:AttachRepositoryNamespace on * for any policy that allows fs:CreateRepository ".  Three possibilities that I can see:
Golang all the way!  Pros: probably more readable, easier to get right. uses a good programming language.  Cons: (very) poor ops experience: not part of the migration library that we use (or, indeed, any reasonable migration library). 
PostgreSQL jsonb SQL whackiness.  Write 100% safe pure-SQL code that updates the table.  Pros: Perfect ops experience.  Cons: The function itself will be tricky to write - it needs to handle IAM-style wildcards (*, ?) entirely correctly.
PostgreSQL jsonb SQL whackiness but less whacky.  Write SQL code that works on every reasonable IAM policy, but fails if the policy uses special (unused!) characters in action names.  Fail migration if such a policy is found (unlikely; such a policy includes an action that is unused!).  Specifically, fail for actions that include SQL-style wildcards (_. %); there are no such actions in lakeFS (or anywhere in IAM, really).

Unless I hear strenuous objections -- hopefully accompanied by suggestions how to do it better -- I am going with the third option.

**Comment by arielshaqed** [comment]
@ozkatz I added data migration; PTAL!  The strategy is non-trivial, so please read also the e8f26a536 commit description.

File: pkg/ddl/000032_iam_add_attach_namespace.up.sql [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    END;
+$$;
+
+-- jsonb_string translates a JSONB string object to SQL TEXT.  See
+-- https://stackoverflow.com/a/58755595.
+CREATE OR REPLACE FUNCTION pg_temp.jsonb_string(j JSONB) RETURNS TEXT LANGUAGE sql IMMUTABLE AS $$
+    SELECT j ->> 0
+$$;
+
+UPDATE auth_policies
```
Comment (ozkatz) [inline pkg/ddl/000032_iam_add_attach_namespace.up.sql:28]: We don't do integration tests for migrations, but I'm assuming this was somehow tested nonetheless?

**Review by ozkatz — APPROVED** [review]
LGTM!

**Review by ozkatz — CHANGES_REQUESTED** [review]
Actually, not LGTM :)
`jsonb_path_exists` is supported on PostgreSQL >= 12
As per [our docs](https://docs.lakefs.io/deploy/aws.html#creating-the-database-on-aws-rds) we support >= 11 (I believe this is also why your tests are failing)

File: pkg/ddl/000032_iam_add_attach_namespace.up.sql (hunk 1 above)
Comment (arielshaqed) [reply pkg/ddl/000032_iam_add_attach_namespace.up.sql:28]: Tested manually: I added manual policies and upgraded and downgraded.  Also upgraded multiple times to ensure idempotency.

You're correct, of course: we have no Postgres testing facility to run unit tests on these migration scripts.

**Review by ozkatz — APPROVED** [review]
Thanks! *Now* it LGTM :)

**Comment by arielshaqed** [comment]
> Thanks! _Now_ it LGTM :)

Thanks!  Also a shout-out to our CI, which came through again :-)

**Pull request merged**